xcm/xcm-executor/src/lib.rs

```diff
@@ -541,6 +556,7 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				// weight consumed correctly (potentially allowing them to do more operations in a
 				// block than they otherwise would).
 				self.total_surplus.saturating_accrue(surplus);
+				RECURSION_COUNT.with(|count| count.set(count.get().saturating_sub(1)));
```

Comment:
DQ but does this not also need to be decremented in the error case?  
(We have a `defer!` macro for that)